presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -408,10 +409,10 @@ public Page getOutput()
             outputIterator = null;
 
             if (finishing) {
-                if (!inputProcessed && !step.isOutputPartial()) {
+                if (!inputProcessed && (step.isOutputPartial() || step.equals(SINGLE))) {
```

Comment:
This file definitely needs more comments - it's convoluted enough that it takes me 30 minutes or so to re-load the design into my head every time I look at it.